concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -1196,15 +1197,69 @@ handleContractUpdateV1 originAddr istance checkAndGetSender transferAmount recei
+                             newBalance WasmV1.Success (Just returnValue))
+                  WasmV1.QueryContractBalance {..}  -> do
+                    newBalance <- getCurrentContractAmount Wasm.SV1 istance
+                    -- TODO Charge query cost or maybe not
+                    -- Lookup contract balances
+                    maybeInstanceInfo <- getCurrentContractInstance imqcbAddress
+                    case maybeInstanceInfo of
+                      Nothing ->
+                        go events =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False
+                             newBalance (WasmV1.Error $ WasmV1.EnvFailure $ WasmV1.MissingContract imqcbAddress) Nothing)
+                      Just instanceInfo -> do
+                        -- Lookup contract balance
+                        balance <- case instanceInfo of
+                                     InstanceInfoV0 _ -> getCurrentContractAmount Wasm.SV0 istance
+                                     InstanceInfoV1 _ -> getCurrentContractAmount Wasm.SV1 istance
+                        -- Construct the return value
+                        let returnValue = WasmV1.byteStringToReturnValue $ S.runPut $ do
+                             S.put balance
+                        go events =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False
+                             newBalance WasmV1.Success (Just returnValue))
+                  WasmV1.QueryExchangeRates -> do
+                    newBalance <- getCurrentContractAmount Wasm.SV1 istance
+                    -- TODO Charge query cost or maybe not
+                    -- Lookup exchange rates
+                    currentExchangeRates <- getExchangeRates
+                    -- Construct the return value
+                    let returnValue = WasmV1.byteStringToReturnValue $ S.runPut $ do
+                          S.put currentExchangeRates
+                    go events =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False
+                          newBalance WasmV1.Success (Just returnValue))
 
       -- start contract execution.
       -- transfer the amount from the sender to the contract at the start. This is so that the contract may immediately use it
       -- for, e.g., forwarding.
       withToContractAmountV1 sender istance transferAmount $ do
         foreignModel <- getRuntimeReprV1 model
         artifact <- liftLocal $ getModuleArtifact (GSWasm.miModule moduleInterface)
-        go [] =<< runInterpreter (return . WasmV1.applyReceiveFun artifact cm receiveCtx receiveName useFallback parameter transferAmount foreignModel)
+        go [] =<< runInterpreter (return . WasmV1.applyReceiveFun artifact cm receiveCtx receiveName useFallback parameter transferAmount foreignModel (supportsChainQueryContracts $ protocolVersion @(MPV m)))
```

Comment:
Consider making a variable for this, so the line doesn't become longer than necessary 😊